torch/distributed/fsdp/_shard_utils.py

```diff
@@ -69,6 +70,10 @@ def _gather_state_dict(
                 tensor = output_tensor.to(local_shard_device)
             else:
                 tensor = output_tensor
+        elif isinstance(tensor, DTensor):
+            tensor = tensor.device_mesh.all_gather(
```

Comment:
I think we should use the `redistribute` API instead of this device_mesh.allgather.